engine/execution/provider/engine_test.go

```diff
@@ -269,38 +313,43 @@ func TestProviderEngine_onChunkDataRequest(t *testing.T) {
 		blockID := unittest.IdentifierFixture()
 
 		execState.On("GetBlockIDByChunkID", chunkID).Return(blockID, nil)
-		ps.On("AtBlockID", blockID).Return(ss)
-
+		ps.On("AtBlockID", blockID).Return(ss).Once()
 		ss.On("Identity", originIdentity.NodeID).Return(originIdentity, nil).Once()
+
+		// channel tracking for the first chunk data pack request responded.
+		responded := make(chan struct{})
```

Comment:
This will work, however the fact that this is necessary in this test case suggests to me that we have a race condition in the other test cases. 

My suggestion to split the "request_shoveler" goroutine from the "request_processor" goroutines was so that we could generally guarantee that a request is processed before we stop the engine, by:
* Call `Process` on some message
* Wait for `engine.MessageStore` to be empty
* Call `cancel` to drain then terminate the engine

I think we are missing the middle step in the testing code:
> Wait for `engine.MessageStore` to be empty

In this test case, the `responded` channel guarantees that the request is processed before the test ends. But we don't have that guarantee for the other test cases. 

If we add this middle step to all test cases (both success and failure), we guarantee the request is processed before the test ends in all cases, removing this race condition. (Adding a concrete suggestion in next comment).

Reply:
https://github.com/onflow/flow-go/pull/2951/commits/f32db517ed868b7636ee1aa577c2d2889d2e6bc9